sshd-core/src/main/java/org/apache/sshd/client/kex/DHGClient.java

```diff
@@ -154,4 +168,54 @@ public boolean next(int cmd, Buffer buffer) throws Exception {
 
         return true;
     }
+
+    protected void verifyCertificate(Session session, OpenSshCertificate openSshKey) throws Exception {
+        PublicKey signatureKey = openSshKey.getCaPubKey();
+        String keyAlg = KeyUtils.getKeyType(signatureKey);
+        Signature verif = ValidateUtils.checkNotNull(
+                NamedFactory.create(session.getSignatureFactories(), keyAlg),
+                "No verifier located for algorithm=%s", keyAlg);
+        verif.initVerifier(session, signatureKey);
+        verif.update(session, openSshKey.getMessage());
+        if (!verif.verify(session, openSshKey.getSignature())) {
+            throw new SshException(SshConstants.SSH2_DISCONNECT_KEY_EXCHANGE_FAILED,
+                    "KeyExchange CA signature verification failed for key type=" + keyAlg);
+        }
+
+        if (openSshKey.getType() != OpenSshCertificate.SSH_CERT_TYPE_HOST) {
+            throw new SshException(SshConstants.SSH2_DISCONNECT_KEY_EXCHANGE_FAILED,
+                    "KeyExchange signature verification failed, not a host key (2): "
+                            + openSshKey.getType());
+        }
+
+        long now = System.currentTimeMillis() / 1000;
+        if (now <= openSshKey.getValidAfter() || now >= openSshKey.getValidBefore()) {
```

Comment:
This condition seems wrong - shouldn't it be the other way around ? I.e., it is valid if `before <= now <= after`

Reply:
>A certificate is considered valid if:
The <= should be a strict <, but with the negation it should be correct besides this.

Reply:
Yes - written this way seems clearer